controller/lib/core/controller.cpp

```diff
@@ -19,65 +19,69 @@ limitations under the License.
 
+static DebugFloat dbg_kp("kp", "Proportional gain for main loop", 3.5);
+static DebugFloat dbg_ki("ki", "Integral gain for main loop", 1.0);
+static DebugFloat dbg_kd("kd", "Derivitive gain for main loop");
+static DebugFloat dbg_sp("pc_setpoint", "Pressure control setpoint (cmH2O)");
+
 Controller::Controller()
-    : pid_(Kp, Ki, Kd, ProportionalTerm::ON_ERROR,
-           DifferentialTerm::ON_MEASUREMENT,
-           // Increases in the blower fan speed should result in increased
-           // pressure.
-           ControlDirection::DIRECT,
-           // Our output is an 8-bit PWM.
-           /*output_min=*/0.f, /*output_max=*/255.f, PID_SAMPLE_PERIOD) {}
+    : pid_(0.0f, 0.0f, 0.0f, ProportionalTerm::ON_ERROR,
+           DifferentialTerm::ON_MEASUREMENT, 0.f, 1.0f, PID_SAMPLE_PERIOD) {}
 
 Duration Controller::GetLoopPeriod() { return PID_SAMPLE_PERIOD; }
 
 ActuatorsState Controller::Run(Time now, const VentParams &params,
                                const SensorReadings &readings) {
   BlowerSystemState desired_state = fsm_.DesiredState(now, params);
 
-  return {.fan_setpoint_cm_h2o = desired_state.setpoint_pressure.cmH2O(),
-          .expire_valve_state = desired_state.expire_valve_state,
-          .fan_power = ComputeFanPower(now, desired_state, readings),
-          .fan_valve = 1.0f /* not used yet */};
-}
+  ActuatorsState actuator_state;
 
-float Controller::ComputeFanPower(Time now,
-                                  const BlowerSystemState &desired_state,
-                                  const SensorReadings &sensor_readings) {
-  // If the blower is not enabled, immediately shut down the fan.  But for
-  // consistency, we still run the PID iteration above.
-  float output;
+  pid_.SetKP(dbg_kp.Get());
+  pid_.SetKI(dbg_ki.Get());
+  pid_.SetKD(dbg_kd.Get());
+
+  // Not used yet
+  actuator_state.fio2_valve = 0.0f;
+
+  float pressure = cmH2O(readings.patient_pressure_cm_h2o).kPa();
   float setpoint = desired_state.setpoint_pressure.kPa();
-  dbg_setpoint.Set(setpoint);
-
-  if (desired_state.blower_enabled) {
-    output = pid_.Compute(
-        /*time=*/now,
-        /*input=*/cmH2O(sensor_readings.patient_pressure_cm_h2o).kPa(),
-        /*setpoint=*/setpoint);
-  } else {
-    output = 0;
-    pid_.Observe(/*time=*/now,
-                 /*input=*/cmH2O(sensor_readings.patient_pressure_cm_h2o).kPa(),
-                 /*setpoint=*/setpoint,
-                 /*output=*/output);
+  dbg_sp.Set(desired_state.setpoint_pressure.cmH2O());
+
+  switch (params.mode) {
+  case VentMode_OFF:
+    actuator_state.setpoint_cm_h2o = 0.0f;
+    actuator_state.blower_valve = std::nullopt;
+    actuator_state.exhale_valve = std::nullopt;
+    actuator_state.blower_power = 0.0f;
+    break;
+
+  case VentMode_PRESSURE_CONTROL:
+
+    // If the pinch valves aren't ready yet then keep the
+    // blower off so we don't force full pressure down the
+    // patients throat.
+    if (!AreActuatorsReady()) {
```

Comment:
Hm can this perhaps be done at a higher level in actuators.cpp, or be passed as an additional input to Controller::Run()? I'm concerned about accessing what looks like hardware-specific code from Controller::Run() which is supposed to be pure math, so that it can be invoked from Modelica.

Reply:
This is a tricky one.  The controller probably needs to have some state logic in it to make sure we move from one state to another in a reasonable way.

Pulling that up out of the switch statement doesn't seem to make sense because we'll need to handle the not ready state differently depending on mode.  When OFF we want to be disabled, when on we don't but need to set the valves to a non-null value to ensure that we are enabling.  We also want logic that controls the blower and other actuators dependent on whether we're enabled or not.

Pulling it right out of stepper.cpp doesn't seem like a good option either because it means that state logic lives both in and out of stepper.cpp.

From a practical standpoint we can probably just have AreActuatorsReady() return true for Modelica and be done with it.

I'll leave this for now and we can discuss further on slack.